rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java

```diff
@@ -142,37 +142,104 @@ public void postMessage(Peer sender, Message message) {
+                allowByMessageCount(sender) && 
+                allowByMinerNotBanned(sender, message) &&
+                allowByMessageUniqueness(sender, message); // prevent repeated is the most expensive and MUST be the last 
+
+    }
+
+    /**
+     * assert score is acceptable 
+     */
+    private boolean allowByScore(double score) {
+        return score >= 0;
+    }
+    
+    /**
+     * assert message count is under the threshold defined in config
+     */
+    private boolean allowByMessageCount(Peer sender) {
+        boolean allow = messageCounter.getValue(sender) < messageQueueMaxSize;
+        if (!allow && logger.isInfoEnabled()) {
+            logger.info("Peer [{}] has its queue full(maxSize: {}). Its messages will not be allowed for a while.", sender.getPeerNodeID(), messageQueueMaxSize);
+        }
+        return allow;
+    }
+
+    private boolean allowByMinerNotBanned(Peer sender, Message message) {
+
+        boolean allow = true;
+        
         if (!this.bannedMiners.isEmpty() && message.getMessageType() == MessageType.BLOCK_MESSAGE) {
             RskAddress miner = ((BlockMessage) message).getBlock().getCoinbase();
             if (this.bannedMiners.contains(miner)) {
                 logger.trace("Received block mined by banned miner {} from peer {}, not added to the queue", miner, sender);
-                return;
+                allow = false;
             }
         }
 
+        return allow;
+    }
+    
+    /**
+     * assert message was not received twice
+     * add it to a map and manages the state of the map
+     * record event if message is repeated 
+     */
+    private boolean allowByMessageUniqueness(Peer sender, Message message) {
+
         Keccak256 encodedMessage = new Keccak256(HashUtil.keccak256(message.getEncoded()));
-        if (!receivedMessages.contains(encodedMessage)) {
+
+        boolean contains = receivedMessages.contains(encodedMessage);
+
+        if (!contains) {
             if (message.getMessageType() == MessageType.BLOCK_MESSAGE || message.getMessageType() == MessageType.TRANSACTIONS) {
                 if (this.receivedMessages.size() >= MAX_NUMBER_OF_MESSAGES_CACHED) {
                     this.receivedMessages.clear();
                 }
                 this.receivedMessages.add(encodedMessage);
             }
 
-            double score = sender.score(System.currentTimeMillis(), message.getMessageType());
-
-            this.addMessage(sender, message, score);
         } else {
             recordEvent(sender, EventType.REPEATED_MESSAGE);
             logger.trace("Received message already known, not added to the queue");
         }
+
+        return !contains;
     }
 
     private void addMessage(Peer sender, Message message, double score) {
-        if (score >= 0 && !this.queue.offer(new MessageTask(sender, message, score))) {
+
+        boolean messageAdded = this.queue.offer(new MessageTask(sender, message, score));
+
+        if (messageAdded) {
```

Comment:
Since `PriorityBlockingQueue` is unbounded `queue.offer` will never return false